go-controller/pkg/node/gateway_localnet.go

```diff
@@ -137,18 +140,23 @@ func (npw *localPortWatcherData) addService(svc *kapi.Service) error {
 				klog.V(5).Infof("ExternalIP: %s is reachable through one of the interfaces on this node, will skip setup", externalIP)
 			} else {
 				if gatewayIP != "" {
-					if stdout, stderr, err := util.RunIP("route", "replace", externalIP, "via", gatewayIP, "dev", localnetGatewayNextHopPort, "table", localnetGatewayExternalIDTable); err != nil {
-						klog.Errorf("Error adding routing table entry for ExternalIP %s: stdout: %s, stderr: %s, err: %v", externalIP, stdout, stderr, err)
-					} else {
-						klog.V(5).Infof("Successfully added route for ExternalIP: %s", externalIP)
-					}
+					routeUsage[externalIP] = true
```

Comment:
I don't fully see the point of this `routeUsage` map. It seems to me like the only concrete change in this file is:  `localnetGatewayNextHopPort` -> `localGwModeNextHopPort`, why change it more than that?

Reply:
its 2 commits. The 2nd commit explains it:
Fixes duplicate route adds/deletes with services

The add/delete route logic was embedded in the for loop on ports and
therefore duplicate adds and deletes were being executed. The extra add
would cause an RTNETLINK no such process error. This patch separates
route operations outside of the ports loop.